File: src/core/api/tauri/system.ts
import { invoke } from '@tauri-apps/api/core'
import type { SystemInfo, AppVersion } from './types'

/**
 * Get system information
 */
export const getSystemInfo = async (): Promise<SystemInfo> => {
  try {
    return await invoke<SystemInfo>('get_system_info')
  } catch (error) {
    console.error('Failed to get system info:', error)
    throw error
  }
}

/**
 * Get application version information
 */
export const getAppVersion = async (): Promise<AppVersion> => {
  try {
    return await invoke<AppVersion>('get_app_version')
  } catch (error) {
    console.error('Failed to get app version:', error)
    throw error
  }
}

/**
 * Check if the application is running in development mode
 */
export const isDevMode = async (): Promise<boolean> => {
  try {
    return await invoke<boolean>('is_dev_mode')
  } catch (error) {
    console.error('Failed to check dev mode:', error)
    throw error
  }
}

/**
 * Get environment variable (filtered for security)
 */
export const getEnvVar = async (key: string): Promise<string | null> => {
  try {
    return await invoke<string | null>('get_env_var', { key })
  } catch (error) {
    console.error(`Failed to get environment variable ${key}:`, error)
    throw error
  }
}

/**
 * Legacy greet function for backward compatibility
 */
export const greet = async (name: string): Promise<string> => {
  try {
    return await invoke<string>('greet', { name })
  } catch (error) {
    console.error('Failed to greet:', error)
    throw error
  }
}

File: src-tauri/src/commands/system.rs
use crate::models::SystemInfo;
use crate::error::AppError;
use serde::{Deserialize, Serialize};
use std::env;

#[derive(Debug, Serialize, Deserialize)]
pub struct AppVersion {
    pub version: String,
    pub build: String,
}

/// Get system information
#[tauri::command]
pub async fn get_system_info() -> Result<SystemInfo, AppError> {
    let os = env::consts::OS;
    let arch = env::consts::ARCH;
    let family = env::consts::FAMILY;
    
    Ok(SystemInfo {
        os: os.to_string(),
        arch: arch.to_string(),
        family: family.to_string(),
        hostname: hostname::get()
            .map_err(|e| AppError::SystemError(format!("Failed to get hostname: {}", e)))?
            .to_string_lossy()
            .to_string(),
    })
}

/// Get application version information
#[tauri::command]
pub async fn get_app_version() -> Result<AppVersion, AppError> {
    Ok(AppVersion {
        version: env!("CARGO_PKG_VERSION").to_string(),
        build: "development".to_string(), // Could be set via build script
    })
}

/// Check if the application is running in development mode
#[tauri::command]
pub async fn is_dev_mode() -> Result<bool, AppError> {
    Ok(cfg!(debug_assertions))
}

/// Get environment variables (filtered for security)
#[tauri::command]
pub async fn get_env_var(key: String) -> Result<Option<String>, AppError> {
    // Only allow specific environment variables for security
    let allowed_vars = vec![
        "NODE_ENV",
        "TAURI_ENV",
        "RUST_LOG",
    ];
    
    if !allowed_vars.contains(&key.as_str()) {
        return Err(AppError::SecurityError(format!("Access to environment variable '{}' is not allowed", key)));
    }
    
    Ok(env::var(&key).ok())
}
